### PR TITLE
Made shutting down nodes non-blocking.

### PR DESCRIPTION
**Pull request by pedrolf**
ptal

### CODE REVIEW

File: rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java [hunk 1]
```diff
@@ -115,9 +124,16 @@ public ScheduledExecutorService getScheduledExecutorService() {
...
   }
 
+  public boolean isShuttingDown() {
+    return shuttingDown;
+  }
+
   @Override
   public void execute(final NodeMain nodeMain, final NodeConfiguration nodeConfiguration,
       final Collection<NodeListener> nodeListeners) {
+    if(isShuttingDown()) {
```
Comment (damonkohler) [inline rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java]: Formatting

File: rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java [hunk 2]
```diff
@@ -115,9 +124,16 @@ public ScheduledExecutorService getScheduledExecutorService() {
...
+      throw new RosRuntimeException("Executing nodes after shutdown has been started.");
```
Comment (damonkohler) [inline rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java]: "Called execute() after calling shutdown()."

File: rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java [hunk 3]
```diff
@@ -115,9 +124,16 @@ public ScheduledExecutorService getScheduledExecutorService() {
     return scheduledExecutorService;
   }
 
+  public boolean isShuttingDown() {
+    return shuttingDown;
```
Comment (damonkohler) [inline rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java]: Instead, just check if shutdownCountDownLatch is null.

Also, need to take the mutex here.

File: rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java (hunk 1 above)
Comment (damonkohler) [inline rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java]: Need to take mutex for this whole method.

File: rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java [hunk 4]
```diff
@@ -137,7 +153,9 @@ public void run() {
         }
         // The new Node will call onStart().
         Node node = nodeFactory.newNode(nodeConfigurationCopy, nodeListenersCopy);
-        nodeMains.put(node, nodeMain);
+        synchronized (mutex) {
```
Comment (damonkohler) [inline rosjava/src/main/java/org/ros/node/DefaultNodeMainExecutor.java]: This runnable needs to call a method on the NodeMainExecutor that takes the mutex. As is, nodeListeners and nodeMains could change while shutting down.

File: rosjava/src/main/java/org/ros/node/NodeMainExecutor.java [hunk 5]
```diff
@@ -62,6 +62,8 @@ void execute(NodeMain nodeMain, NodeConfiguration nodeConfiguration,
    */
   void execute(NodeMain nodeMain, NodeConfiguration nodeConfiguration);
 
+  void addListener(NodeMainExecutorListener listener);
```
Comment (damonkohler) [inline rosjava/src/main/java/org/ros/node/NodeMainExecutor.java:71]: Javadoc

**Comment by pedrolf** [comment]
I only changed NodeMainExecutor.java and NodeMainExecutorListener.java adding the missing comments.

**Comment by damonkohler** [comment]
LGTM

**Pull request merged**